### ISSUE TITLE
[Payment due 2024-09-09][$250] [Xero] Tapping "Other integrations" in accounting page slight flashing / janky animation

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#43397]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.81-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @shawnborton 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1718032595682989

## Action Performed:
Have a QBO connection
1. Open app
2. Go to workspace
3. Go to accounting
4. Tap " Other integrations"
## Expected Result:
There should be smooth animation
## Actual Result:
There is a slight flashing/janky animation
## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [x] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/38435837/11478888-3152-4781-8904-cc48f40cbad2

https://github.com/Expensify/App/assets/38435837/596bb9bc-8801-4bdc-8b56-9a2d69ce75a6

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01894cdf15ac4a5e92</li>
        <li>Upwork Job ID: 1800299476976101068</li>
        <li>Last Price Increase: 2024-06-17</li>
<li>Automatic offers: </li>
<ul>
<li>Pujan92 | Reviewer | 102770646</li>
<li>truph01 | Contributor | 102770647</li>
</ul></li>
    </ul>
</details>

**Comment by trjExpensify** [comment]
CC: @rushatgabhane @mananjadhav @hungvu193 seen this before?

**Comment by rushatgabhane** [comment]
@trjExpensify i was testing this and i noticed a bug on android - https://expensify.slack.com/archives/C049HHMV9SM/p1718039521119029

Touch inputs don't work when signing in to quickbooks / xero

**Comment by rushatgabhane** [comment]
@trjExpensify haven't seen before but i can reproduce this bug

**Comment by trjExpensify** [comment]
Let's send it external then!

**Comment by truph01** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
There is a slight flashing/janky animation of the collapse component

### What is the root cause of that problem?
The bug is in `react-native-collapsible`

There's a condition mismatch between https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L188 and https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L194.

The `hasKnownHeight` is used to render the parent view with fixed height or not. So when `hasKnownHeight` is `false` initially when `collapsed` becomes `false`, there will be no style [here](https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L226). At the same time, the measuring doesn't start yet, so the `measuring` [here](https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L194) is `false`, so the content becomes rendered.

Right after that the `measuring` becomes `true` and the content [becomes hidden](https://github.com/oblador/react-native-collapsible/blob/master/Collapsible.js#L195-L196), then after measuring completed, the content becomes visible again. 

So the root cause is that although the `hasKnownHeight` is `false` (the parent view is not ready to be shown yet), we're still showing the content (because measuring didn't start).

### What changes do you think we should make in order to solve the problem?

Make the condition between https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L188 and https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L194 consistent.

If `hasKnownHeight` is false (the parent view is [not ready to be shown yet](https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L189)), the content should also not show.

[This condition](https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L194) can be updated to:
```
if (!hasKnownHeight) {
```
The `!hasKnownHeight` condition is inclusive of the existing `(measuring)` condition because if `measuring` is `true`, `hasKnownHeight` will be `false` and `!hasKnownHeight` will also be `true`. So this will not cause any problem.

### What alternative solutions did you explore? (Optional)
NA

**Comment by truph01** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
There is a slight flashing/janky animation of the collapse component

### What is the root cause of that problem?
The bug is in `react-native-collapsible`

There's a condition mismatch between https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L188 and https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L194.

The `hasKnownHeight` is used to render the parent view with fixed height or not. So when `hasKnownHeight` is `false` initially when `collapsed` becomes `false`, there will be no style [here](https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L226). At the same time, the measuring doesn't start yet, so the `measuring` [here](https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L194) is `false`, so the content becomes rendered.

Right after that the `measuring` becomes `true` and the content [becomes hidden](https://github.com/oblador/react-native-collapsible/blob/master/Collapsible.js#L195-L196), then after measuring completed, the content becomes visible again. 

So the root cause is that although the `hasKnownHeight` is `false` (the parent view is not ready to be shown yet), we're still showing the content (because measuring didn't start).

### What changes do you think we should make in order to solve the problem?

Make the condition between https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L188 and https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L194 consistent.

If `hasKnownHeight` is false (the parent view is [not ready to be shown yet](https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L189)), the content should also not show.

[This condition](https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L194) can be updated to:
```
if (!hasKnownHeight) {
```
The `!hasKnownHeight` condition is inclusive of the existing `(measuring)` condition because if `measuring` is `true`, `hasKnownHeight` will be `false` and `!hasKnownHeight` will also be `true`. So this will not cause any problem.

### What alternative solutions did you explore? (Optional)
Update the logic here https://github.com/oblador/react-native-collapsible/blob/2bd7f8d50c2f3d876119bc65da8e7681c38563ad/Collapsible.js#L189-L192 so if `hasKnownHeight` is `false`, the parent view will have invisible style
```
const style = hasKnownHeight ? {
      overflow: 'hidden',
      height: height,
    } : {
      overflow: 'hidden',
      height: 0,
    };
```

Or just 
```
const style = {
      overflow: 'hidden',
      height: height,
    };
```
As when `hasKnownHeight` is `false`, the `height` will also be 0

**Comment by TheGithubDev** [comment]
### **Proposal**

### **Please re-state the problem that we are trying to solve in this issue:**
The problem is that when a user navigates to "Other integrations" in the accounting section of the workspace, the animation is not smooth. Instead, there is a slight flashing animation.

### **What is the root cause of that problem?**
The root cause of the janky animation could be due to:
1. **Inefficient Rendering:** Inefficient rendering or re-rendering of components causing the animation to lag.
2. **Heavy Components:** Heavy components or operations being loaded synchronously, which blocks the main thread and causes animation to stutter.
3. **CSS/Style Issues:** Improper use of CSS or style properties that interfere with smooth animations.

**What changes do you think we should make in order to solve the problem?**
To address this issue, we need to:
1. Optimize Rendering:
   - Ensure that components are rendered efficiently, possibly using React.memo or PureComponent to avoid unnecessary re-renders.
2. Asynchronous Loading:
   - Load heavy components or data asynchronously to prevent blocking the main thread during animations.
3. CSS/Style Improvements:
   - Review and optimize CSS or style properties used for animations to ensure smooth transitions.

### **Example of changes:**

1. ### **Optimize Rendering:**
   - Use React.memo or PureComponent to prevent unnecessary re-renders.
   ```javascript
   const OtherIntegrations = React.memo(() => {
       return (
           <View>
               {/* Other integration components */}
           </View>
       );
   });
   ```

2. ### **Asynchronous Loading:**
   - Load heavy components or data asynchronously.
   ```javascript
   useEffect(() => {
       const loadHeavyComponents = async () => {
           await loadData();
       };

       loadHeavyComponents();
   }, []);
   ```

3. ### **CSS/Style Improvements:**
   - Ensure CSS animations are optimized and using properties that can be hardware accelerated, like transform and opacity.
   ```css
   .integration-animation {
       transition: transform 0.3s ease-in-out, opacity 0.3s ease-in-out;
   }
   ```

### **Alternative solutions ?**
1. **Performance Profiling**: Use performance profiling tools to identify and address specific bottlenecks causing the animation to jank.
2. **Virtualization**: Implement virtualization for lists or heavy components to load only visible items, reducing the rendering load.
3. **Animation Libraries**: Utilize animation libraries like Lottie or React Native Reanimated to handle complex animations more efficiently.

_Please note that the above explanation outlines only the technical approach I plan to take. If my proposal gets accepted, I will submit the finalized solution, as mentioned._

_Thanks!_

**Comment by lschurr** [comment]
@Pujan92 can you take a look at the proposals here?

**Comment by Pujan92** [comment]
As @truph01's mentioned in their [proposal](https://github.com/Expensify/App/issues/43397#issuecomment-2159823409) the bug seems to be in `react-native-collapsible`. A similar issue https://github.com/oblador/react-native-collapsible/issues/473 is raised in the same repo.

I agree with @truph01's RCA where due to those conditions the content gets rendered before the parent View style has been assigned. I like their alternative solution to provide `overflow: 'hidden'` in each cases for the parent to avoid overflowing the content. I believe `hasKnownHeight` part can be omitted.
For this we need to raise a PR in `react-native-collapsible` and I think the reviewer will help there in case `hasKnownHeight` is required.

🎀👀🎀 C+ reviewed

**Comment by lschurr** [comment]
@neil-marcellini - can you take a quick look at this one?

**Comment by neil-marcellini** [comment]
> As @truph01's mentioned in their [proposal](https://github.com/Expensify/App/issues/43397#issuecomment-2159823409) the bug seems to be in `react-native-collapsible`. A similar issue [oblador/react-native-collapsible#473](https://github.com/oblador/react-native-collapsible/issues/473) is raised in the same repo.
> 
> I agree with @truph01's RCA where due to those conditions the content gets rendered before the parent View style has been assigned. I like their alternative solution to provide `overflow: 'hidden'` in each cases for the parent to avoid overflowing the content. I believe `hasKnownHeight` part can be omitted. For this we need to raise a PR in `react-native-collapsible` and I think the reviewer will help there in case `hasKnownHeight` is required.
> 
> 🎀👀🎀 C+ reviewed

Sounds good, hiring @truph01

**Comment by truph01** [comment]
@Pujan92 With this issue, should I create a pull request (PR) to fix the library directly or create a patch package?

**Comment by Pujan92** [comment]
I prefer to create PR in the lib but since long the repo isn't been updated, so we can proceed with the patch package option.
WDYT @neil-marcellini ?

**Comment by rushatgabhane** [comment]
@Pujan92 i think it's best that we discuss on opensource channel. We have precedents for this

**Comment by Pujan92** [comment]
Thanks @rushatgabhane, I will check and raise it on the open-source channel.

**Comment by Pujan92** [comment]
@truph01 Based on the discussion [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1718735142327679), let's raise a PR in the library repo and we try to inform the maintainer to review it.

**Comment by lschurr** [comment]
Any update on this one @truph01?

**Comment by truph01** [comment]
@lschurr I created PR in that lib https://github.com/oblador/react-native-collapsible/pull/475

**Comment by Pujan92** [comment]
Thanks @truph01.
On Twitter, they have responded to me and I have shared the PR link. They will check the PR after their vacation.
Let's wait for some days.

**Comment by Pujan92** [comment]
Waiting for a lib PR to be reviewed.

**Comment by truph01** [comment]
@Pujan92 Do you know when the lib's author can review the PR?

**Comment by Pujan92** [comment]
I sent a bump message yesterday, we will ask in the slack thread if we don't get a response untill Monday.

**Comment by lschurr** [comment]
Any update @Pujan92?

**Comment by Pujan92** [comment]
@lschurr We are still waiting on the review of that lib PR by the maintainer, I will ask @AndrewGable in slack in case they can help to inform the author for the re-review.

**Comment by AndrewGable** [comment]
I believe the PR was reviewed, but it wasn't working as intended in the test app, so I would confirm the fix there.

**Comment by Pujan92** [comment]
@lschurr The other issue https://github.com/Expensify/App/issues/44264 needs to be put on Hold not this one.

@ishpaul Yes, it seems the same issue. We are waiting for the lib PR to re-review. meanwhile, you can give that [PR](https://github.com/oblador/react-native-collapsible/pull/475) a look if you want.

**Comment by truph01** [comment]
@Pujan92 Can you ping the author to review the PR?

**Comment by Pujan92** [comment]
I just did and sent them a message on Twitter. Let's see.

**Comment by truph01** [comment]
@Pujan92 Do you have any updates here?

**Comment by Pujan92** [comment]
I got a response on 25th July -- **I am, but on another business trip 🙈 Would it help if I pushed a branch where I reproduce the blinking?**

I have requested them to share the branch once they push, but seems they are quite busy so let's wait.

**Comment by truph01** [comment]
@Pujan92 Did the author respond?

**Comment by truph01** [comment]
@Pujan92 If the author is not available to review the PR, we can consider creating a patch. What do you think?

**Comment by Pujan92** [comment]
Yes, we can consider that option as PR review in lib is taking a long. WDYT @neil-marcellini?

**Comment by truph01** [comment]
What do you think about [this idea](https://github.com/Expensify/App/issues/43397#issuecomment-2301870945) @neil-marcellini ?

**Comment by truph01** [comment]
Still waiting for a response from @neil-marcellini.

**Comment by neil-marcellini** [comment]
Yeah let's do a patch since it's been a while and the maintainer is unresponsive, please go ahead and create a PR.

**Comment by Pujan92** [comment]
Lib [PR](https://github.com/oblador/react-native-collapsible/pull/475) gets merged, @truph01 We can raise a PR with a version bump.

**Comment by Pujan92** [comment]
@truph01 Seems our changes are causing an issue https://github.com/oblador/react-native-collapsible/issues/479 where the content isn't rendered if the initial value for the expanded is true, Let's try to find and fix it.

**Comment by truph01** [comment]
@Pujan92 I am investigating the RCA

**Comment by truph01** [comment]
Update: I am trying to find the RCA. Also, I tried applying other solutions mentioned in [here](https://github.com/Expensify/App/issues/43397#issuecomment-2159823409) but still encountered different problems.

**Comment by truph01** [comment]
PR was deployed to production on 31/8/2024

**Comment by lschurr** [comment]
Ok, thanks! Just clarifying, [this PR](https://github.com/Expensify/App/pull/48145) fixed the bug right? And then we take [this other one](https://github.com/Expensify/App/issues/44264) off hold?

**Comment by lschurr** [comment]
Payment summary:

- Contributor: @truph01 $250 ([paid in Upwork](https://www.upwork.com/nx/wm/workroom/37697818/overview))
- Reviewer: @Pujan92 $250 (please request in NewDot)

**Comment by truph01** [comment]
I just want to point out: To fix [this issue](https://github.com/Expensify/App/issues/43397), we need to submit a PR to the [react-native-collapsible](https://github.com/oblador/react-native-collapsible) library. Once that PR is merged, our Expensify app work correctly. However, the [react-native-collapsible](https://github.com/oblador/react-native-collapsible) library encountered a new [issue](https://github.com/oblador/react-native-collapsible/issues/479), and we are still investigating the root cause along with other users.

**Comment by garrettmknight** [comment]
$250 approved for @Pujan92
